src/Adapter/MSTest.TestAdapter/Helpers/MethodRunner.cs

```diff
@@ -0,0 +1,54 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Globalization;
+using System.Reflection;
+
+using Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.ObjectModel;
+
+using UnitTestOutcome = Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.ObjectModel.UnitTestOutcome;
+
+namespace Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.Helpers;
+
+internal static class MethodRunner
+{
+    internal static TestFailedException? RunWithTimeoutAndCancellation(
+        Action action, CancellationTokenSource cancellationTokenSource, int? timeout, MethodInfo methodInfo,
+        string methodCancelledMessageFormat, string methodTimedOutMessageFormat)
+    {
+        try
+        {
+            var executionTask = Task.Run(action, cancellationTokenSource.Token);
```

Comment:
this can throw `TaskCanceledException` that's `OperationCanceledException` but not a timeout or do we fold every `OperationCanceledException` of this method in a "timeout"?

Reply:
Actually this one doesn't throw OCE according to doc